compiler/rustc_mir_dataflow/src/value_analysis.rs

```diff
@@ -0,0 +1,940 @@
+/// the bottom element (because joining an unreachable and any other reachable state yields a
+/// reachable state). All operations on unreachable states are ignored.
+///
+/// Flooding means assigning a value (by default `⊤`) to all tracked projections of a given place.
+#[derive(PartialEq, Eq, Debug)]
+pub struct State<V>(StateData<V>);
+
+impl<V: Clone> Clone for State<V> {
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+
+    fn clone_from(&mut self, source: &Self) {
+        self.0.clone_from(&source.0);
+    }
+}
+
+impl<V: Clone + HasTop + HasBottom> State<V> {
+    pub fn is_reachable(&self) -> bool {
+        matches!(&self.0, StateData::Reachable(_))
+    }
+
+    pub fn mark_unreachable(&mut self) {
+        self.0 = StateData::Unreachable;
+    }
+
+    pub fn flood_all(&mut self) {
+        self.flood_all_with(V::top())
+    }
+
+    pub fn flood_all_with(&mut self, value: V) {
+        let StateData::Reachable(values) = &mut self.0 else { return };
+        values.raw.fill(value);
+    }
+
+    pub fn flood_with(&mut self, place: PlaceRef<'_>, map: &Map, value: V) {
+        if let Some(root) = map.find(place) {
+            self.flood_idx_with(root, map, value);
+        }
+    }
+
+    pub fn flood(&mut self, place: PlaceRef<'_>, map: &Map) {
+        self.flood_with(place, map, V::top())
+    }
+
+    pub fn flood_idx_with(&mut self, place: PlaceIndex, map: &Map, value: V) {
+        let StateData::Reachable(values) = &mut self.0 else { return };
+        map.preorder_invoke(place, &mut |place| {
+            if let Some(vi) = map.places[place].value_index {
+                values[vi] = value.clone();
+            }
+        });
+    }
+
+    pub fn flood_idx(&mut self, place: PlaceIndex, map: &Map) {
+        self.flood_idx_with(place, map, V::top())
+    }
+
+    /// Copies `source` to `target`, including all tracked places beneath.
+    ///
+    /// If `target` contains a place that is not contained in `source`, it will be overwritten with
+    /// Top. Also, because this will copy all entries one after another, it may only be used for
+    /// places that are non-overlapping or identical.
+    pub fn assign_place_idx(&mut self, target: PlaceIndex, source: PlaceIndex, map: &Map) {
+        let StateData::Reachable(values) = &mut self.0 else { return };
+
+        // If both places are tracked, we copy the value to the target. If the target is tracked,
+        // but the source is not, we have to invalidate the value in target. If the target is not
+        // tracked, then we don't have to do anything.
+        if let Some(target_value) = map.places[target].value_index {
+            if let Some(source_value) = map.places[source].value_index {
+                values[target_value] = values[source_value].clone();
+            } else {
+                values[target_value] = V::top();
+            }
+        }
+        for target_child in map.children(target) {
```

Comment:
A small question. What is the protocol for such parameters?
- Have they to be constants, with names and documentation? Or just a value and a comment pointing out what the value's role is?
- They should be tuned at some point, right? To get the best performance possible out of the pass. Is this to be added as an issue later?

Reply:
There is no policy surrounding this as far as I am aware. Below is my opinion.

> Have they to be constants, with names and documentation? Or just a value and a comment pointing out what the value's role is?

I think I would prefer a constant from a code review perspective. At the very least it makes it easier to keep track of the value as it moves around in other code changes.

> They should be tuned at some point, right? To get the best performance possible out of the pass. Is this to be added as an issue later?

Tuning these parameters is a positive contribution. That being said, I can think of at least two reasons that doing this may at times have limited value:

 1. If there are future changes planned to the pass, those changes might invalidate any previous tuning.
 2. Other changes might be a more time efficient investment.

I suspect that both of these are the case here. As a result, I don't think filing an issue makes too much sense, especially since we won't be filing issues for any of the other outstanding improvements (maybe we should?). If one day we get to the point where we feel this is some form of "done," that calculus might change.

However, it seems reasonable to include a comment for these parameters indicating the degree to which the choice of value is evidence based vs a guess (and a short description of how to reproduce any evidence would be great too) - @jachris can you add this?

Reply:
Mm, this makes sense to me :D thanks for the answers!

Regarding tuning... perhaps at some point parameters like these (in the entirety of the compiler) could be automatically tuned before a new beta is branched off. I can imagine of a way to tune them. It'd be slow, but at least it'd be automatic.

I feel like that + tracking over time of the value of these tunings could provide interesting improvements and insights over the compiler's performance.

It's no small job, though. Maybe someday when I'm better at Rust I'll take it up :3

Reply:
> could be automatically tuned before a new beta is branched off

In principle yes, although you would have a pretty hard time selling me on any system to do it (don't let that stop you from trying though :P )

Reply:
> Fyi this is the part that I was talking about before, where it seems like we're doing unnecessary work in the case of a type like `[u8; 1000]`. Not to be fixed in this PR obviously, but with some smart engineering it feels like we could do a much better job here

I see what you mean. Perhaps one of the requirements for the improvements after this PR should be optimize moves. This can increase both performance and accuracy. (Btw, the elements of `[u8; 1000]` will not be tracked currently, but a struct with 1000 fields would have the same issue.)